terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,239 @@
+func resourceRoleRead(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
-	return nil
+	r, err := getRole(ctx, d, c)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	if r == nil {
+		return diag.Diagnostics{}
+	}
+
+	err = tfschema.SetRoleV3(r, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diag.Diagnostics{}
 }
 
-func resourceTeleportRoleDelete(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleUpdate updates Teleport role from resource definition
+func resourceRoleUpdate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
 
-	return nil
+	r, err := getRole(ctx, d, c)
+	if err != nil {
+		return diag.FromErr(describeErr(err, "role"))
+	}
+	if r == nil {
+		return diag.Diagnostics{}
+	}
+
+	removeMapAndListDefaults(r)
+
+	err = tfschema.GetRoleV3(r, d)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = c.UpsertRole(ctx, r)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return resourceRoleRead(ctx, d, m)
+}
+
+// resourceRoleDelete deletes Teleport role from resource definition
+func resourceRoleDelete(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	id := d.Id()
+	err = c.DeleteRole(ctx, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return diag.Diagnostics{}
 }
 
-func resourceTeleportRoleExists(d *schema.ResourceData, m interface{}) (bool, error) {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// getRole gets role with graceful destroy handling
+func getRole(ctx context.Context, d *schema.ResourceData, c *client.Client) (*types.RoleV3, error) {
+	id := d.Id()
+
+	r, err := c.GetRole(ctx, id)
+	if trace.IsNotFound(err) {
+		d.SetId("")
+		return nil, nil
+	}
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
 
-	return true, nil
+	r3, ok := r.(*types.RoleV3)
+	if !ok {
+		return nil, fmt.Errorf("failed to convert created user to types.RoleV3 from %T", r)
+	}
+
+	removeMapAndListDefaults(r3)
+
+	return r3, nil
+}
+
+// removeComplexDefaults removes invalid default items
+func removeMapAndListDefaults(r *types.RoleV3) {
+	removeDefaultLabel(r.Spec.Allow.AppLabels)
+	removeDefaultLabel(r.Spec.Allow.DatabaseLabels)
+	removeDefaultLabel(r.Spec.Allow.NodeLabels)
+	removeDefaultLabel(r.Spec.Allow.KubernetesLabels)
+	removeDefaultLabel(r.Spec.Deny.AppLabels)
+	removeDefaultLabel(r.Spec.Deny.DatabaseLabels)
+	removeDefaultLabel(r.Spec.Deny.NodeLabels)
+	removeDefaultLabel(r.Spec.Deny.KubernetesLabels)
+
+	if reflect.DeepEqual(r.Spec.Options.BPF, defaultBPF) {
+		r.Spec.Options.BPF = []string{}
+	}
+}
+
+// removeDefaultLabel removes label "*":"*" which is mistakenly assigned on Teleport side
```

Comment:
Does this really need to be worked-around in the terraform provider? We should (and will) fix it in the product itself. I'm just wondering about potential subtle issues if this will remove something that it's not supposed to.

Reply:
Unfortunately, there's no other way for now. If user won't specify *:* in his Terraform config, these fields will be set immediately after object creation and will be written to the local state. And this will cause state drift on next apply. We can not make them computed either because Terraform support for complex computed fields is broken.